File: reactive-web/src/main/java/com/github/corneil/reactivespring5/service/LocationHistoryService.java
package com.github.corneil.reactivespring5.service;

import com.github.corneil.reactivespring5.model.ExtendedLocationHistory;
import com.github.corneil.reactivespring5.model.LocationHistory;
import com.github.corneil.reactivespring5.repository.LocationHistoryRepository;
import lombok.extern.slf4j.XSlf4j;
import org.springframework.data.mongodb.core.geo.GeoJsonPoint;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Date;

@Service("locationHistory")
@XSlf4j
public class LocationHistoryService implements LocationHistoryInterface {
	private LocationHistoryRepository historyRepository;

	public LocationHistoryService(LocationHistoryRepository historyRepository) {
		this.historyRepository = historyRepository;
	}

	@Override
	public Flux<LocationHistory> findByDates(Date startDate, Date endDate) {
		try {
			log.info(">>findByDates");
			return historyRepository.findByTimestampBetween(startDate, endDate);
		} finally {
			log.info("<<findByDates");
		}
	}

	@Override
	public Flux<ExtendedLocationHistory> findAndConvert(Date startDate, Date endDate) {
		try {
			log.info(">>findAndConvert");
			return historyRepository.findByTimestampBetween(startDate, endDate)
				.flatMap(item -> Mono.just(convertToExtended(item)));
		} finally {
			log.info("<<findAndConvert");
		}
	}

	private static ExtendedLocationHistory convertToExtended(LocationHistory item) {
		final double x = item.getLocation().getX();
		final double y = item.getLocation().getY();
		return new ExtendedLocationHistory(item.getId(),
			item.getTimestamp(),
			new GeoJsonPoint(y, x),
			String.format("Location %f, %f", x, y));
	}
}

File: reactive-kotlin/src/main/kotlin/com/github/corneil/reactivespring5kotlin/service/LocationHistoryService.kt
package com.github.corneil.reactivespring5kotlin.service

import com.github.corneil.reactivespring5kotlin.model.ExtendedLocationHistory
import com.github.corneil.reactivespring5kotlin.model.LocationHistory
import com.github.corneil.reactivespring5kotlin.repository.LocationHistoryRepository
import mu.KLogging
import org.springframework.data.mongodb.core.geo.GeoJsonPoint
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.util.*

@Service("locationHistory")
class LocationHistoryService(val historyRepository: LocationHistoryRepository) : LocationHistoryInterface {
	companion object : KLogging()

	override fun findByDates(startDate: Date, endDate: Date): Flux<LocationHistory> {
		try {
			logger.info(">>findByDates")
			return historyRepository.findByTimestampBetween(startDate, endDate)
		} finally {
			logger.info("<<findByDates")
		}
	}

	override fun findAndConvert(startDate: Date, endDate: Date): Flux<ExtendedLocationHistory> {
		try {
			logger.info(">>findAndConvert")
			return historyRepository.findByTimestampBetween(startDate, endDate).flatMap { item -> Mono.just(convertToExtended(item)) }
		} finally {
			logger.info("<<findAndConvert")
		}
	}

	private fun convertToExtended(item: LocationHistory): ExtendedLocationHistory {
		val x = item.location.x
		val y = item.location.y
		return ExtendedLocationHistory(item.id,
			item.timestamp,
			GeoJsonPoint(y, x),
			String.format("Location %f, %f", x, y))
	}
}
